File: library/src/main/java/com/lazygeniouz/house/ads/helper/RemoveJsonObjectCompat.java
/*
 * Created by Darshan Pandya.
 * @itznotabug
 * Copyright (c) 2018.
 */

package com.lazygeniouz.house.ads.helper;

import android.os.AsyncTask;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

public class RemoveJsonObjectCompat extends AsyncTask<JSONArray, JSONArray, JSONArray> {
    private JSONArray jsonArray;
    private int index;

    public RemoveJsonObjectCompat(int index, JSONArray array) {
        this.jsonArray = array;
        this.index = index;
    }


    @Override
    protected JSONArray doInBackground(JSONArray... jsonArrays) {
        return removeJsonObject(index, jsonArray);
    }


    private JSONArray removeJsonObject(final int idx, final JSONArray from) {
        final List<JSONObject> objects = asList(from);
        objects.remove(idx);

        final JSONArray ja = new JSONArray();
        for (final JSONObject obj : objects) {
            ja.put(obj);
        }

        return ja;
    }

    private List<JSONObject> asList(final JSONArray ja) {
        final int len = ja.length();
        final ArrayList<JSONObject> result = new ArrayList<>(len);
        for (int i = 0; i < len; i++) {
            final JSONObject obj = ja.optJSONObject(i);
            if (obj != null) {
                result.add(obj);
            }
        }
        return result;
    }

}

File: library/src/main/java/com/lazygeniouz/house/ads/helper/RemoveJsonObjectCompat.kt
/*
 * Created by Darshan Pandya.
 * @itznotabug
 * Copyright (c) 2018.
 */

package com.lazygeniouz.house.ads.helper

import android.os.AsyncTask
import androidx.annotation.RestrictTo
import org.json.JSONArray
import org.json.JSONObject
import java.util.*

/*
`JsonArray.remove(index)` is only available on API 19+,
so this utility class helps us do that dirty work on pre 19.
 */

@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
class RemoveJsonObjectCompat(private val index: Int, private val jsonArray: JSONArray) : AsyncTask<JSONArray, JSONArray, JSONArray>() {

    override fun doInBackground(vararg jsonArrays: JSONArray): JSONArray {
        return removeJsonObject(index, jsonArray)
    }

    private fun removeJsonObject(child: Int, array: JSONArray): JSONArray {
        val objects = asList(array)
        objects.removeAt(child)

        val jsonArray = JSONArray()
        for (obj in objects) jsonArray.put(obj)
        return jsonArray
    }

    private fun asList(jsonArray: JSONArray): MutableList<JSONObject> {
        val length = jsonArray.length()
        val result = ArrayList<JSONObject>(length)
        for (index in 0 until length) {
            val element = jsonArray.optJSONObject(index)
            if (element != null) result.add(element)
        }
        return result
    }
}